### PR TITLE
Export from riak to zip

### PR DESCRIPTION
**Pull request by emord**
This uses the existing migration framework to export a domain's attachments (except sql forms for now).

It uses a management command for importing from zip to an existing riak, because there will be a lot of buildup for it to be the input to the migration framework for something that will likely only be used once (progress loader, document provider, document processor...)

@millerdev @snopoke @czue

### CODE REVIEW

File: corehq/blobs/zipdb.py
Comment (emord) [inline corehq/blobs/zipdb.py]: Need to read the whole file in memory because of limitations with `zipfile`

File: corehq/blobs/__init__.py
Comment (millerdev) [inline corehq/blobs/__init__.py]: I'm pretty sure this should not be using `MigratingBlobDB` since the only operation called on this db is `copy_blob`, which should only affect the zip db.

File: corehq/blobs/migrate.py
Comment (millerdev) [inline corehq/blobs/migrate.py]: Might be worth writing a new document provider that iterates over documents in the domain rather than all documents of the given types. Could also make a provider that can handle types from multiple couch dbs to allow a single export with all blobs rather than one for each couch db. Might also need a different progress logger implementation to give accurate progress output.

File: corehq/blobs/zipdb.py
Comment (millerdev) [inline corehq/blobs/zipdb.py]: I think this should `raise NotImplementedError`; should only be using `copy_blob`.

File: corehq/blobs/zipdb.py
Comment (millerdev) [inline corehq/blobs/zipdb.py]: What do you think of opening the zip file here in `__init__` and adding a `close` method (called by `BlobDbBackendExporter.processing_complete`) rather than open/close on each `copy_blob`. Seems like it might be useful to check if the export file already exists at the beginning of the export and refuse to modify it. Could add an `--append` flag to the management command if you want to override that sanity check, but I assume that wouldn't be the default case.

Also, do you think the export and import management commands should document in their help that they use this filename (in the current directory)? Might also be nice to add an option to specify this file path on the command line.

File: corehq/blobs/zipdb.py
Comment (millerdev) [inline corehq/blobs/zipdb.py]: This looks the same as the version in fsdb. Why not import `safejoin` from there?

File: corehq/blobs/migrate.py
Comment (millerdev) [inline corehq/blobs/migrate.py]: Why not set `self.db = get_blob_db_exporter(slug, domain)` in `__init__` above?

File: corehq/blobs/migrate.py
Comment (millerdev) [inline corehq/blobs/migrate.py]: This class should override `_backup_doc` with a no-op. I think there's no need to write anything to the migration log for exports (especially not a full json dump of the doc).

**Review by millerdev — CHANGES_REQUESTED** [review]
Nice! There are a few things that need some attention yet, but this is looking good.

What's your plan for new attachments that are written to the db between when the migration starts and when the restore happens?

**Comment by millerdev** [comment]
I see there are no tests. Do you think this code is code will be used infrequently enough that it's not worth checking that it actually reads and writes the correct content?

File: corehq/blobs/management/commands/import_blob_zip.py
Comment (emord) [inline corehq/blobs/management/commands/import_blob_zip.py]: Yeah. You can't use `seek` in a `ZipFileExt`. I also looked at using tarfiles, but there are similar issues there

File: corehq/blobs/migrate.py
Comment (dimagimon) [inline corehq/blobs/migrate.py]: F401 'memoized' imported but unused

File: corehq/blobs/__init__.py
Comment (emord) [inline corehq/blobs/__init__.py]: https://github.com/dimagi/commcare-hq/pull/13478/commits/25637ae268f334eeb2d7aafafddec9ae9f89a186

File: corehq/blobs/zipdb.py
Comment (emord) [inline corehq/blobs/zipdb.py]: https://github.com/dimagi/commcare-hq/pull/13478/commits/25637ae268f334eeb2d7aafafddec9ae9f89a186

File: corehq/blobs/management/commands/import_blob_zip.py
Comment (emord) [inline corehq/blobs/management/commands/import_blob_zip.py]: https://github.com/dimagi/commcare-hq/pull/13478/commits/f5f6d4abad7c58eb92d6de11b666f9e3e94c1a9e

File: corehq/blobs/migrate.py
Comment (emord) [inline corehq/blobs/migrate.py]: https://github.com/dimagi/commcare-hq/pull/13478/commits/1000633a123822ae4b9c79e20799b6bb45a18a3e

File: corehq/blobs/__init__.py
Comment (emord) [inline corehq/blobs/__init__.py]: https://github.com/dimagi/commcare-hq/pull/13478/commits/55f06adcea7fa97659a32b0dde38e5e52a4abf06 & https://github.com/dimagi/commcare-hq/pull/13478/commits/b8ff4f485d46c608e28a6e71a8678797c78a6fde

I'm fine with either approach, but I'm sure why one would be better than hte other. I originally used `MigratingBlobDB` because you need access to both the zipfile "db" and the current riak instance. It does seem weird because we're not "migrating" though.

File: corehq/blobs/migrate.py
Comment (emord) [inline corehq/blobs/migrate.py]: That was because I was inheriting the exporter from the migrator which would check the db against `MigratingBlobDB`. changed in  55f06ad & b8ff4f4

File: corehq/blobs/zipdb.py
Comment (emord) [inline corehq/blobs/zipdb.py]: I did it this way, so that the chances of a corrupt zip file are very small in comparison to having a file open for a very long time. zipfiles don't provide any sort of `flush` command

Comment (emord) [inline corehq/blobs/zipdb.py]: agreed about documentation and extra flags. can do that more easily in a follow up PR

**Comment by emord** [comment]
> What's your plan for new attachments that are written to the db between when the migration starts and when the restore happens?

I'm not worried about this because for the icds migration we are stopping submissions and the app is frozen. Should revisit if this code is used again

File: corehq/blobs/migrate.py
Comment (emord) [inline corehq/blobs/migrate.py]: Interesting idea. I had not thought about having a `DomainDocProvider`

**Comment by snopoke** [comment]
LGTM assuming there will be a followup PR to address some of the comments. @millerdev waiting for your approval.

File: corehq/blobs/migrate.py
Comment (emord) [inline corehq/blobs/migrate.py]: Ah thanks.I had started looking at implementing that myself. I don't think case attachments are a thing that is deployed yet, so probably won't need the case provider right?

File: corehq/blobs/migrate.py
Comment (dimagimon) [inline corehq/blobs/migrate.py]: E226 missing whitespace around arithmetic operator

File: corehq/blobs/migrate.py
Comment (millerdev) [inline corehq/blobs/migrate.py]: What do you think of moving this string to a global constant since it's used in a number of places (DRY)?

File: corehq/blobs/zipdb.py
Comment (millerdev) [inline corehq/blobs/zipdb.py]: nit: could combine above two functions into
```py
def get_blob_db_exporter(slug, domain):
    return ZipBlobDB(slug, domain)
```

**Review by millerdev — CHANGES_REQUESTED** [review]
Looking good. I think there's still one broken thing (noted in review comment).

File: corehq/blobs/migrate.py
Comment (snopoke) [inline corehq/blobs/migrate.py]: right

File: corehq/blobs/zipdb.py
Comment (emord) [inline corehq/blobs/zipdb.py]: Thanks that's good to know. I was just looking at the zip file api. I'd be hesitant to rely on private implementation

**Pull request merged**
